primitives/runtime/src/transaction_validity.rs

```diff
@@ -161,6 +161,35 @@ impl Into<TransactionValidity> for UnknownTransaction {
 	}
 }
 
+/// The origin of the transaction.
+///
+/// Depending on the source we might apply different validation schemes.
+/// For instance we can disallow specific kinds of transactions if they were not produced
+/// by our local node (for instance off-chain workers).
+#[derive(Copy, Clone, PartialEq, Eq, Encode, Decode, RuntimeDebug, parity_util_mem::MallocSizeOf)]
+pub enum TransactionSource {
+	/// Transaction is already included in block.
+	///
+	/// This means that we can't really tell where the transaction is coming from,
+	/// since it's already in the received block. Note that the custom validator
+	/// using either `Local` or `External` should most likely just allow `InBlock`
+	/// transactions as well.
+	InBlock,
```

Comment:
A few notes about this: 

>  Note that the custom **validator**.. 

I think what you mean here is a _custom unsigned validation_? now it confuses me with the term validator as in _validator node_.

> since it's already in the received block

Am I correct to think that also when you create a block locally, once you are done with validating all the transactions, in the execution phase, you will see this `InBlock`? Of course all other nodes who receive the block will also see `InBlock`

Reply:
If I am correct, then maybe the other two variants can be called `ValidationLocal` and `ValidationExternal` since you basically only see them in the authoring's validation phase. Once you are passed this, everyone will see it as `InBlock `.

Reply:
We use `validator` in a sense of `UnsignedValidator` logic in that file already, but I rephrased to "custom validation logic" to make it clear.

> Am I correct to think that also when you create a block locally, once you are done with validating all the transactions, in the execution phase, you will see this InBlock? Of course all other nodes who receive the block will also see InBlock

For in-block transactions we don't really run `UnsginedValidator::validate` at all, cause we run `pre_dispatch` instead. However the default `pre_dispatch` implementation just delegates to `validate` so I decided to call this case `InBlock`.

The only case we ever see `Local/External` is when we validate for transaction pool, block execution will always see only `InBlock` (due to `pre_dispatch`, not an explicity `validate` call).